src/shred_network/duplicate_shred_listener.zig

```diff
@@ -0,0 +1,794 @@
+    epoch_ctx_mgr: *sig.adapter.EpochContextManager,
+};
+
+pub const DuplicateShredListener = struct {
+    thread: std.Thread,
+
+    pub fn start(
+        allocator: Allocator,
+        logger: Logger,
+        params: Params,
+    ) !DuplicateShredListener {
+        const thread = try std.Thread.spawn(
+            .{},
+            recvLoop,
+            .{ allocator, logger, params },
+        );
+        errdefer thread.join();
+
+        thread.setName("sigCiEntryLstnr") catch {};
+        return .{ .thread = thread };
+    }
+
+    pub fn join(self: DuplicateShredListener) void {
+        self.thread.join();
+    }
+};
+
+const Key = struct { slot: Slot, from: Pubkey };
+
+const BufferEntry = struct {
+    chunks: [MAX_NUM_CHUNKS]?[]u8,
+};
+
+pub fn recvLoop(
+    allocator: Allocator,
+    logger: Logger,
+    params: Params,
+) !void {
+    var handler = DuplicateShredHandler.init(allocator, logger, params);
+    defer handler.deinit();
+
+    var cursor: usize = 0;
+    var buf: [512]GossipVersionedData = undefined;
+
+    while (!params.exit.load(.monotonic)) {
+        const gossip_data: []GossipVersionedData = blk: {
+            const gossip_table, var lock = params.gossip_table_rw.readWithLock();
+            defer lock.unlock();
+            break :blk gossip_table.getClonedEntriesWithCursor(
+                allocator,
+                &buf,
+                &cursor,
+            ) catch |e| {
+                logger.err().logf("duplicate_shred_listener: get entries failed: {}", .{e});
+                break :blk &.{};
+            };
+        };
+
+        if (gossip_data.len == 0) {
+            std.time.sleep(GOSSIP_SLEEP_MILLIS * std.time.ns_per_ms);
+            continue;
+        }
+
+        for (gossip_data) |*versioned| {
+            defer versioned.deinit(allocator);
+            switch (versioned.data) {
+                .DuplicateShred => |data| {
+                    _, const dup = data;
+                    handler.handle(dup) catch |e| handler.logger.err().logf(
+                        "duplicate_shred_listener: handle chunk failed for slot {}: {}",
+                        .{ dup.slot, e },
+                    );
+                },
+                else => {},
+            }
+        }
+    }
+}
+
+const DuplicateShredHandler = struct {
+    allocator: Allocator,
+    logger: Logger,
+    params: Params,
+    // Because we use UDP for packet transfer, we can normally only send ~1500 bytes
+    // in each packet. We send both shreds and meta data in duplicate shred proof, and
+    // each shred is normally 1 packet(1500 bytes), so the whole proof is larger than
+    // 1 packet and it needs to be cut down as chunks for transfer. So we need to piece
+    // together the chunks into the original proof before anything useful is done.
+    dup_buffer: std.AutoHashMapUnmanaged(Key, BufferEntry),
+    consumed: std.AutoHashMapUnmanaged(Slot, bool),
+    last_root: Slot,
+    cached_slots_in_epoch: u64,
```

Comment:
iiuc `consumed` and `last_root` are also "cached" state, yeah? If so, would be nice to annotate that, preferably by reflecting that in their names - at the very least a minimal high level description of how they are updated would be appreciated.